File: python/producer/producer.py
import requests
import json
import time
import random
from datetime import datetime
from typing import Dict, Any
import argparse
from requests.exceptions import RequestException
import sys

# API details
URL = "https://mpk-8-gateway.staging-cloud.nussknacker.io/topics/http.tests"
USERNAME = "publisher"
PASSWORD = "QTbp6MQL1bvn9unVrg0XzsXEfff"

SAMPLE_DATA = {
    "names": ["Alice", "Bob", "Charlie", "David", "Eve", "Frank", "Grace", "Hannah"],
    "cities": ["New York", "London", "Tokyo", "Paris", "Berlin", "Sydney"],
    "products": ["Laptop", "Phone", "Tablet", "Watch", "Headphones"],
    "statuses": ["pending", "completed", "failed", "in_progress"]
}

# Template for the message structure
# Modify this dictionary to change the structure of your messages
# MESSAGE_TEMPLATE = {
#     "name": "random_name",
#     "city": "random_city",
# }

MESSAGE_TEMPLATE = {
    "user": {
        "name": "random_name",  # Will be replaced with random name
        "city": "random_city"   # Will be replaced with random city
    },
    "order": {
        "product": "random_product",
        "quantity": "random_int(1,5)",
        "status": "random_status",
        "timestamp": "current_timestamp"
    }
}

class ValidationError(Exception):
    pass

def get_random_value(field_type):
    """Generate random values based on the field type"""
    if field_type == "random_name":
        return random.choice(SAMPLE_DATA["names"])
    elif field_type == "random_city":
        return random.choice(SAMPLE_DATA["cities"])
    elif field_type == "random_product":
        return random.choice(SAMPLE_DATA["products"])
    elif field_type == "random_status":
        return random.choice(SAMPLE_DATA["statuses"])
    elif field_type == "current_timestamp":
        return datetime.now().isoformat()
    elif field_type.startswith("random_int"):
        # Parse the range from the string "random_int(min,max)"
        min_val, max_val = map(int, field_type[11:-1].split(','))
        return random.randint(min_val, max_val)
    return field_type

def generate_data(template=MESSAGE_TEMPLATE):
    """Recursively generate data based on the template"""
    if isinstance(template, dict):
        return {k: generate_data(v) for k, v in template.items()}
    elif isinstance(template, list):
        return [generate_data(item) for item in template]
    elif isinstance(template, str):
        return get_random_value(template)
    return template

def send_data():
    data = generate_data()
    headers = {"Content-Type": "application/json"}
    
    try:
        response = requests.post(
            URL, 
            auth=(USERNAME, PASSWORD), 
            headers=headers, 
            data=json.dumps(data)
        )

        if response.status_code == 400 and "Invalid message" in response.text:
            raise ValidationError(
                f"\n\nSchema validation error detected!\n"
                f"Your message structure might not match the expected Avro schema.\n\n"
                f"Current message structure:\n{json.dumps(data, indent=2)}\n\n"
                f"Suggested Avro schema for your current message structure:\n"
                f"{generate_avro_schema()}\n\n"
                f"Please update the schema in Nu Cloud to match your message structure.\n"
                f"You can copy the schema above and use it in the Nu Cloud interface.\n"
            )
        response.raise_for_status()
        print(f"✅ Sent message: {data}")
    except ValidationError as e:
        print(str(e))
        sys.exit(1)
    except RequestException as e:
        print(f"❌ Failed to send: {data}, Response: {e}")

def infer_avro_type(value: str) -> Dict[str, Any]:
    """Infer Avro type from template value"""
    if value == "random_name" or value == "random_city" or value == "random_product" or value == "random_status":
        return {"type": "string"}
    elif value == "current_timestamp":
        return {"type": "string", "logicalType": "iso-datetime"}
    elif value.startswith("random_int"):
        return {"type": "int"}
    return {"type": "string"}  # default fallback

def generate_avro_schema(template: Dict = MESSAGE_TEMPLATE, name: str = "Message") -> str:
    """Generate Avro schema from message template"""
    def _process_field(value: Any, field_name: str) -> Dict[str, Any]:
        if isinstance(value, dict):
            return {
                "name": field_name,
                "type": {
                    "type": "record",
                    "name": field_name.capitalize(),
                    "fields": [
                        _process_field(v, k) 
                        for k, v in value.items()
                    ]
                }
            }
        elif isinstance(value, str):
            field_schema = infer_avro_type(value)
            return {"name": field_name, **field_schema}
        else:
            return {"name": field_name, "type": "string"}  # fallback

    schema = {
        "type": "record",
        "name": name,
        "namespace": "com.example",
        "fields": [
            _process_field(value, key)
            for key, value in template.items()
        ]
    }
    
    return json.dumps(schema, indent=2)

if __name__ == "__main__":
    parser = argparse.ArgumentParser(description='Message producer with schema generation capability')
    parser.add_argument('--schema', action='store_true', 
                       help='Generate and print Avro schema instead of producing messages')
    args = parser.parse_args()

    if args.schema:
        print("Generated Avro Schema:")
        print(generate_avro_schema())
    else:
        try:
            while True:
                send_data()
                time.sleep(1)
        except KeyboardInterrupt:
            print("\nStopping message production.")
            sys.exit(0)

File: node.js/producer/producer.js
const axios = require('axios');
const yaml = require('js-yaml');
const fs = require('fs');
const path = require('path');

// Sample data for random generation
const SAMPLE_DATA = {
    names: ["Alice", "Bob", "Charlie", "David", "Eve", "Frank", "Grace", "Hannah"],
    cities: ["New York", "London", "Tokyo", "Paris", "Berlin", "Sydney"],
    products: ["Laptop", "Phone", "Tablet", "Watch", "Headphones"],
    statuses: ["pending", "completed", "failed", "in_progress"]
};

// Template for the message structure
const MESSAGE_TEMPLATE = {
    "name": "random_name",
};

const EXAMPLE_MESSAGE_TEMPLATE = {
    user: {
        name: "random_name",
        city: "random_city"
    },
    order: {
        product: "random_product",
        quantity: "random_int(1,5)",
        status: "random_status",
        timestamp: "current_timestamp"
    }
};

class ValidationError extends Error {
    constructor(message) {
        super(message);
        this.name = 'ValidationError';
    }
}

function getRandomValue(fieldType) {
    if (fieldType === "random_name") {
        return SAMPLE_DATA.names[Math.floor(Math.random() * SAMPLE_DATA.names.length)];
    } else if (fieldType === "random_city") {
        return SAMPLE_DATA.cities[Math.floor(Math.random() * SAMPLE_DATA.cities.length)];
    } else if (fieldType === "random_product") {
        return SAMPLE_DATA.products[Math.floor(Math.random() * SAMPLE_DATA.products.length)];
    } else if (fieldType === "random_status") {
        return SAMPLE_DATA.statuses[Math.floor(Math.random() * SAMPLE_DATA.statuses.length)];
    } else if (fieldType === "current_timestamp") {
        return new Date().toISOString();
    } else if (fieldType.startsWith("random_int")) {
        const [min, max] = fieldType.slice(11, -1).split(',').map(Number);
        return Math.floor(Math.random() * (max - min + 1)) + min;
    }
    return fieldType;
}

function generateData(template = MESSAGE_TEMPLATE) {
    if (typeof template === 'object' && template !== null) {
        return Object.fromEntries(
            Object.entries(template).map(([k, v]) => [k, generateData(v)])
        );
    } else if (Array.isArray(template)) {
        return template.map(item => generateData(item));
    } else if (typeof template === 'string') {
        return getRandomValue(template);
    }
    return template;
}

function inferAvroType(value) {
    if (["random_name", "random_city", "random_product", "random_status"].includes(value)) {
        return { type: "string" };
    } else if (value === "current_timestamp") {
        return { type: "string", logicalType: "iso-datetime" };
    } else if (value.startsWith("random_int")) {
        return { type: "int" };
    }
    return { type: "string" };
}

function generateAvroSchema(template = MESSAGE_TEMPLATE, name = "Message") {
    function processField(value, fieldName) {
        if (typeof value === 'object' && value !== null) {
            return {
                name: fieldName,
                type: {
                    type: "record",
                    name: fieldName.charAt(0).toUpperCase() + fieldName.slice(1),
                    fields: Object.entries(value).map(([k, v]) => processField(v, k))
                }
            };
        } else if (typeof value === 'string') {
            const fieldSchema = inferAvroType(value);
            return { name: fieldName, ...fieldSchema };
        }
        return { name: fieldName, type: "string" };
    }

    const schema = {
        type: "record",
        name: name,
        namespace: "com.example",
        fields: Object.entries(template).map(([key, value]) => processField(value, key))
    };

    return JSON.stringify(schema, null, 2);
}

function loadConfig() {
    const configPath = path.join(__dirname, 'config.yaml');
    const templatePath = path.join(__dirname, 'config.yaml.template');

    if (!fs.existsSync(configPath)) {
        console.log("📝 Creating initial config.yaml from template...");
        fs.copyFileSync(templatePath, configPath);
        console.log("✅ Created config.yaml");
        console.log("⚠️  Please edit config.yaml with your actual configuration values");
        process.exit(1);
    }

    return yaml.load(fs.readFileSync(configPath, 'utf8'));
}

async function sendData(url, username, password, data = null) {
    if (!data) {
        data = generateData();
    }

    try {
        await axios.post(url, data, {
            auth: {
                username: username,
                password: password
            },
            headers: {
                'Content-Type': 'application/json'
            }
        });
        console.log(`✅ Sent message: ${JSON.stringify(data)}`);
    } catch (error) {
        if (error.response?.status === 400 && error.response?.data?.includes("Invalid message")) {
            throw new ValidationError(
                `\n\nSchema validation error detected!\n` +
                `Your message structure might not match the expected Avro schema.\n\n` +
                `Current message structure:\n${JSON.stringify(data, null, 2)}\n\n` +
                `Suggested Avro schema for your current message structure:\n` +
                `${generateAvroSchema()}\n\n` +
                `Please use the topic with the appropriate schema in Nu Cloud matching your message structure.\n` +
                `You can copy the schema above and use it in the Nu Cloud interface to create new topic.\n`
            );
        }
        console.error(`❌ Failed to send: ${JSON.stringify(data)}, Response: ${error.message}`);
    }
}

async function main() {
    const args = process.argv.slice(2);
    if (args.includes('--schema')) {
        console.log("Generated Avro Schema:");
        console.log(generateAvroSchema());
        return;
    }

    const config = loadConfig();
    const { url, username, password } = config.api;
    const { delay_seconds } = config.producer;

    try {
        while (true) {
            await sendData(url, username, password);
            await new Promise(resolve => setTimeout(resolve, delay_seconds * 1000));
        }
    } catch (error) {
        if (error instanceof ValidationError) {
            console.error(error.message);
            process.exit(1);
        }
        throw error;
    }
}

if (require.main === module) {
    process.on('SIGINT', () => {
        console.log("\nStopping message production.");
        process.exit(0);
    });

    main().catch(error => {
        console.error(error);
        process.exit(1);
    });
}
